### PR TITLE
Add dependencies for terminations

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
Add dependencies for terminations

### CODE REVIEW

File: wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/ServiceTerminator.java [hunk 1]
```diff
@@ -30,12 +38,14 @@
...
-  private final Set<ServiceTermination> terminations;
+  private final List<ServiceTermination> terminations;
 
   @Inject
   ServiceTerminator(EventService eventService, Set<ServiceTermination> terminations) {
     this.eventService = eventService;
-    this.terminations = terminations;
+    ArrayList<ServiceTermination> terminationsList = new ArrayList<>(terminations);
+    Collections.sort(terminationsList, new ServiceTerminationComparator(terminations));
+    this.terminations = terminationsList;
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/ServiceTerminator.java]: can we do that in one shot? With stream for example?

File: wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/ServiceTerminator.java [hunk 2]
```diff
@@ -88,4 +99,45 @@ private void doTerminate(ServiceTermination termination) throws InterruptedExcep
...
+    }
+
+    private final Map<String, Set<String>> dependencies;
+
+    @Override
+    public int compare(ServiceTermination o1, ServiceTermination o2) {
+
+      return checkTransitiveDependency(o1.getServiceName(), o2.getServiceName(), new HashSet<>());
+    }
+
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/ServiceTerminator.java]: can you add some info what is going on here?

File: wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/ServiceTerminator.java (hunk 1 above)
Comment (akorneta) [reply wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/ServiceTerminator.java]: +1 consider to do it like `this.terminations = ImmutableSortedSet.copyOf(new Comparator<..>(), terminations);`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesClientTermination.java [hunk 3]
```diff
@@ -48,4 +50,9 @@ public void suspend() throws InterruptedException {
   public String getServiceName() {
     return "KubernetesClient";
   }
+
+  @Override
+  public Set<String> getDependencies() {
+    return ImmutableSet.of("workspace");
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesClientTermination.java]: Please consider using constant from WorkspaceServiceTermination here

**Review by akorneta — APPROVED** [review]

**Review by sleshchenko — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**